### DISCUSSION TITLE
How to use Kerberos

### DISCUSSION THREAD

**Discussion by igh-gr** [FreeRDP/FreeRDP#8553] [Q&A]
Hello,
In the thread about proper kerberos support (https://github.com/FreeRDP/FreeRDP/issues/5746) it says 

> we have full functional Kerberos support

Now I'm not sure about what 'full functional Kerberos support' means. Does that include Single Sign-on support ?

Could you possibly explain how to use the new functionality ?

Sorry if this is already answered somewhere, but I can't find any information on this.

**Comment by akallabeth** [comment]
@igh-gr it means you can now log in with `NLA` and `kerberos`, e.g. use your smartcard to login.
there is currently limited support for autoconfiguration of realms though, so a 'Single sign on' is not yet available (at least not without configuring the connection first)

**Reply by hardening** [reply]
@igh build FreeRDP with kerberos and recommanded SPNego support, and then you'll be able to logon with NLA to a windows server with NTLM auth disabled.
About SSO, if you already have a ccache we will probably use it, but the client will still require that you provide user/password (that's a limit in our implementation right now)

**Reply by YugoCode** [reply]
@hardening, @akallabeth 

> About SSO, if you already have a ccache we will probably use it, but the client will still require that you provide user/password (that's a limit in our implementation right now)

Is this still the case? Or is SSO without providing creds but solely the .ccache file possible?
Thanks!

**Reply by akallabeth** [reply]
@YugoCode @hardening did address this issue recently with https://github.com/FreeRDP/FreeRDP/pull/10549
so now just add `/remoteGuard` and it will reuse your existing kerberos `.ccache`

it might still not work if there is no ticket already (work in progress) but it is a start

**Reply by AkechiShiro** [reply]
Thanks for mentioning this update @akallabeth 

Is there any way to know when the next release will happen, I believe the roadmap is close to 100% for 3.9.0, just would like to test this feature `/remoteGuard` with a ticket provided in `KRB5CCACHE`, if I really want to try it out before the release there is also the option to compile it I guess

**Reply by M4rt1n12** [reply]
Hi, I connected with kinit to the server and can see my connection with klist. If I connect to my server with:
`xfreerdp3 /auth-pkg-list:'!ntlm,kerberos' /u:xxxxxxxxx /p:xxxxxxxxx /v:192.168.30.21 /d:xxxxxxxxx.LAN /cert:ignore`
I get this error: 
`InitializeSecurityContext failed with SEC_E_INTERNAL_ERROR [0x80090304]
[transport_connect_nla]: ERRCONNECT_AUTHENTICATION_FAILED [0x00020009]
`
I'm absolutely sure that the authentication is correct. But I disabled NTLMv1/v2 on this server, so the connection must be established via kerberos.

Maybe I missed a configuration, but I couldn't find a really good guide on this topic. Thanks!

**Reply by Cyb3rC3lt** [reply]
Hi, just getting back to a previous question. Can xfreerdp3 now be used from Linux (Kali) with a ccache in klist without a password or krb5.conf entry? If so can you provide an example command as I see kerberos mentioned in the help and nla etc? 

In earlier posts I also see people saying look for GSSAPI in build config to make sure it's supported but I guess that may not be required anymore if Kerberos is enabled by default? Thanks

**Comment by hardening** [comment]
I'm not sure remote credential guard is the answer to SSO with kerberos as it requires quite a lot of server setup to work.

To do a complete answer (perhaps we should have a Wiki page for that given the number of requests on that subject). Yes we can achieve the NLA auth part when using Kerberos (as negotiated by SPNego) just from the ccache, BUT to complete NLA, we must send a `TSPasswordCreds` packet with user/domain/password to have SSO on the server, so we need the password. If you use the restricted admin flag these fields can be empty, but of course you loose SSO on the server.

As a more general thought, if we had a `lsass4nonwin32` daemon running aside with FreeRDP, we could communicate with that daemon for credentials and access to caches, just like does the lsass process under windows (we could even send NTLM hashes in RCG). Well more precisely we could ask that daemon to do the NLA part like does the tspkg SSPI module on windows with mstsc, and on the FreeRDP side we would have the exact same implementation (so asking to regular LSASS or to `lsass4nonwin32`).

**Comment by opoplawski** [comment]
I'm afraid I don't understand the guts of the protocol here, but I will say that with rdesktop I'm able to do `rdesktop -r scard -p - -o sc-csp-name="Microsoft Base Smart Card Crypto Provider"  -u USER@DOMAIN -o sc-container-name="te-Yubikey Template-b6a101d0-bb01-13144" HOST` then enter my pin and get directly connected to the remote machine without entering my password if I have a kerberos ticket.  So I don't understand the need to enter a password.  This is with NLA security.

**Reply by hardening** [reply]
Well smartcard logon is a bit different, and you can achieve the same with FreeRDP with `/smartcard-logon`.

About details with kerberos: yes you can pass the pure Kerberos steps of NLA with an existing ccache, but until you specify the restricted admin option, that allows to not provide the user's password in clear, you need the password to fill `TSUserPasswordCred` in NLA.

**Reply by hardening** [reply]
BTW one could arg that you should also be able to smartcard logon with a pre-established ccache `kinit -X X509_user_identity='PKCS11:opensc-pkcs11.so' ...`, but in this case the pin is required anyway to fill `TSSmartcardCreds`
